src/index.cpp

```diff
@@ -587,48 +530,15 @@ void Index<T, TagT, LabelT>::load(const char *filename, uint32_t num_threads, ui
 #ifndef EXEC_ENV_OLS
     if (file_exists(labels_file))
     {
-        _label_map = load_label_map(labels_map_file);
-        parse_label_file(labels_file, label_num_pts);
-        assert(label_num_pts == data_file_num_pts - _num_frozen_pts);
-        if (file_exists(labels_to_medoids))
+        if (_filter_store == nullptr)
         {
-            std::ifstream medoid_stream(labels_to_medoids);
-            std::string line, token;
-            uint32_t line_cnt = 0;
-
-            _label_to_start_id.clear();
-
-            while (std::getline(medoid_stream, line))
-            {
-                std::istringstream iss(line);
-                uint32_t cnt = 0;
-                uint32_t medoid = 0;
-                LabelT label;
-                while (std::getline(iss, token, ','))
-                {
-                    token.erase(std::remove(token.begin(), token.end(), '\n'), token.end());
-                    token.erase(std::remove(token.begin(), token.end(), '\r'), token.end());
-                    LabelT token_as_num = (LabelT)std::stoul(token);
-                    if (cnt == 0)
-                        label = token_as_num;
-                    else
-                        medoid = token_as_num;
-                    cnt++;
-                }
-                _label_to_start_id[label] = medoid;
-                line_cnt++;
-            }
-        }
-
-        std::string universal_label_file(filename);
-        universal_label_file += "_universal_label.txt";
-        if (file_exists(universal_label_file))
-        {
-            std::ifstream universal_label_reader(universal_label_file);
-            universal_label_reader >> _universal_label;
-            _use_universal_label = true;
-            universal_label_reader.close();
+            _filter_store = std::make_unique<InMemFilterStore<LabelT>>(_max_points + _num_frozen_pts);
         }
+        _filter_store->load_label_map(labels_map_file);
+        label_num_pts = _filter_store->load_labels(labels_file);
+        assert(label_num_pts == data_file_num_pts);
+        _filter_store->load_medoids(labels_to_medoids);
```

Comment:
Same here. I think single load function we can expose from filter_store and then internally that will load labels, medoids, map and universal label